cfgrammar/src/lib/yacc/parser.rs

```diff
@@ -474,17 +460,59 @@ impl YaccParser {
             i = k;
         }
 
+        // Next, the '(' pattern : type, ... ')'
         i = self.parse_ws(i, false)?;
-        let (i, pat) = self.parse_to_single_colon(i)?;
-        let (i, typ) = self.parse_to_eol(i + ':'.len_utf8())?;
-        let pat = pat.trim_start().trim_end().to_string();
-        let typ = typ.trim_start().trim_end().to_string();
-        if !(pat.is_empty() || typ.is_empty()) {
-            self.ast.parse_param_bindings = Some((pat, typ));
+        if let Some(mut j) = self.lookahead_is("(", i) {
+            let mut bindings: Vec<(String, String)> = Vec::new();
+
+            while j < self.src.len() {
```

Comment:
If we make this something like:

```rust
while j < self.src.len() && self.src[j..].chars().next().unwrap() != ')'
```

then we can get rid of the slightly magical-seeming `done` variable I think?

Reply:
Its a bit difficult to do, because `j` is the character *after* ')'.
I did an update, it used a mutable to put it at the beginning of the loop, but between the break and the mutation I thought it was a toss-up, returning the char we can at least give it a better name.

Reply:
So, I think the last thing I pushed is the closest we can come to that.

Reply:
I wonder if:

```rust
while j < self.src.len() && !self.lookahead_is(")", j) {
   ...
}
```

works? I must admit I haven't thought about this part of the code in ages!

Reply:
It's a bit awkward, i.e. we still need to increment in the loop, and outside the loop with `parse_to_eol()`, but I made it so!